### PR TITLE
receiver: Add a prototype VM metrics receiver.

### PR DESCRIPTION
**Pull request by songy23**
Fixes https://github.com/census-instrumentation/opencensus-service/issues/514.

Only for POC purpose. Expose process and system telemetry and let users have access to it.

vm-metrics-receiver

### CODE REVIEW

File: receiver/vmmetricsreceiver/transform_stats_to_metrics.go [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vmmetricsreceiver
+
+// TODO(songy23): remove this file and use Metrics Gauge and Cumulative APIs.
```
Comment (songy23) [inline receiver/vmmetricsreceiver/transform_stats_to_metrics.go]: This file is a copy of https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/blob/master/transform_stats_to_metrics.go. We will remove this once we migrate to use the new Cumulative and Gauge APIs in Go.

File: receiver/vmmetricsreceiver/transform_stats_to_metrics.go [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+	}
+
+	if len(timeseries) == 0 {
+		return nil, nil
+	}
+
+	return timeseries, nil
+}
+
+func timeToProtoTimestamp(t time.Time) *timestamp.Timestamp {
```
Comment (pjanotti) [inline receiver/vmmetricsreceiver/transform_stats_to_metrics.go]: nit: iirc internal has a helper function for this

Comment (songy23) [reply receiver/vmmetricsreceiver/transform_stats_to_metrics.go]: That's right, updated to use internal.TimeToTimestamp.

File: receiver/vmmetricsreceiver/vm_metrics_collector.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		metricPrefix:   mprefix,
+		done:           make(chan struct{}),
+	}
+	view.Register(vmc.views...)
+	return vmc, nil
+}
+
+// StartCollection starts a ticker'd goroutine that will scrape and export vm metrics periodically.
+func (vmc *VMMetricsCollector) StartCollection() {
+	ticker := time.NewTicker(vmc.scrapeInterval)
```
[the remark this answers is not in the document]
Comment (songy23) [reply receiver/vmmetricsreceiver/vm_metrics_collector.go]: Fixed.

File: receiver/vmmetricsreceiver/vm_metrics_collector.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	}()
+}
+
+// StopCollection stops the collection of metric information
+func (vmc *VMMetricsCollector) StopCollection() {
+	close(vmc.done)
+}
+
+func (vmc *VMMetricsCollector) scrape() {
+	ms := &runtime.MemStats{}
```
[the remark this answers is not in the document]
Comment (songy23) [reply receiver/vmmetricsreceiver/vm_metrics_collector.go]: Exactly, I actually copied the process metrics from internal process_telemetry. Filed https://github.com/census-instrumentation/opencensus-service/issues/518 to keep track of the progress.

File: receiver/vmmetricsreceiver/vm_metrics_collector.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	pid := os.Getpid()
+	proc, err := procfs.NewProc(pid)
+	if err == nil {
+		if procStat, err := proc.NewStat(); err == nil {
+			stats.Record(ctx, mCPUSeconds.M(int64(procStat.CPUTime())))
+		}
+	}
+
+	if stat, err := vmc.fs.NewStat(); err == nil {
+		stats.Record(ctx, mProcessesCreated.M(int64(stat.ProcessCreated)))
```
[the remark this answers is not in the document]
Comment (songy23) [reply receiver/vmmetricsreceiver/vm_metrics_collector.go]: Great catch, actually batch recording of measurement has a better performance. Updated.

File: receiver/vmmetricsreceiver/vm_metrics_collector.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		stats.Record(ctx, mUserCPUSeconds.M(cpuStat.User))
+		stats.Record(ctx, mNiceCPUSeconds.M(cpuStat.Nice))
+		stats.Record(ctx, mSystemCPUSeconds.M(cpuStat.System))
+		stats.Record(ctx, mIdleCPUSeconds.M(cpuStat.Idle))
+		stats.Record(ctx, mIowaitCPUSeconds.M(cpuStat.Iowait))
+	}
+}
+
+func (vmc *VMMetricsCollector) export() {
+	vds := []*view.Data{}
```
[the remark this answers is not in the document]
Comment (songy23) [reply receiver/vmmetricsreceiver/vm_metrics_collector.go]: Agree, updated.

File: receiver/vmmetricsreceiver/vm_metrics_collector.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	ctx, span := trace.StartSpan(context.Background(), "VMMetricsCollector.uploadViewData")
+	defer span.End()
+
+	metrics := make([]*metricspb.Metric, 0, len(vds))
+	for _, vd := range vds {
+		if metric, err := viewDataToMetric(vd); err == nil {
+			metrics = append(metrics, metric)
+		}
```
[the remark this answers is not in the document]
Comment (songy23) [reply receiver/vmmetricsreceiver/vm_metrics_collector.go]: SG, updated.

**Comment by songy23** [comment]
Thanks!

**Pull request merged**